File: 01_binary_search/go/main.go
package main

import (
	"fmt"
)

func main() {
	var list = []int{1, 3, 5, 7, 9}
	fmt.Printf("test 1 %v\n", binary_search(list, 3))
	fmt.Printf("test 2 %v\n", binary_search(list, -1))
}

func binary_search(list []int, item int) int {

	low := 0
	high := len(list) - 1

	for low <= high {
		mid := (low + high) / 2
		guess := list[mid]

		if guess == item {
			return mid
		}

		if guess > item {
			high = mid - 1
		} else {
			low = mid + 1
		}

	}
	return -1
}

File: 01_binary_search/rs/main.rs
fn main() {
    let list: [i32; 5] = [1, 3, 5, 7, 9];
    assert_eq!(binary_search(&list, 3), 1);
    assert_eq!(binary_search(&list, -1), -1);
    println!("test 1 {}", binary_search(&list, 3));
    println!("test 2 {}", binary_search(&list, -1));
}

fn binary_search(list: &[i32], item: i32) -> i32 {
    let mut low: i32 = 0;
    let mut high: i32 = list.len() as i32 - 1;
    while low <= high {
        let mid: i32 = (low + high) / 2;
        let guess: i32 = list[(mid as usize)];
        if guess == item {
            return mid as i32;
        }

        if guess > item {
            high = mid - 1;
        } else {
            low = mid + 1;
        }

    }
    return -1;
}